### ISSUE TITLE
deserialize: flexible defaults for missing struct fields

### ISSUE THREAD

**Issue by anguslees** [serde-rs/serde#90]
Currently I have to define a newtype (eg: "BoolDefaultTrue") every time I want an optional struct field to have a non-Default::default() default, which is clumsy.  Ideally I could just say something like:

```rust
struct Foo {
    #[serde(default=true)]
    enabled: bool,
}
```

... and this would fill in the value `true` if it was missing from the serialised string.  Super-awesome would be if serde also generated a `Default` trait implementation for `Foo` that respected the same values, but I could understand that being considered out of scope.

**Comment by oli-obk** [comment]
If you create a manual `Default` impl for `Foo`, serde still uses the `Default` impl of the fields` type.

Example: this impl doesn't cause the `#[serde(default)]` to set the field to `true`.

```rust
impl Default for Foo {
    fn default() -> Foo {
        Foo {
            enabled: true,
        }
    }
}
```

@erickt: is using the surrounding type's `Default` impl (if present) too much magic? Alternatively requiring that the type impls `Default` if any field is annotated would be a breaking change and maybe not what one expects when annotating a field.

**Comment by erickt** [comment]
This was implemented in #238!

**Comment by dtolnay** [comment]
Yep, this is now doable with the `#[serde(default="...")]` attribute. I think this issue can be closed.
